sources/web/datalab/server.ts

```diff
@@ -250,6 +256,22 @@ function uncheckedRequestHandler(request: http.ServerRequest, response: http.Ser
 // The path that is used for the optional websocket proxy for HTTP requests.
 const httpOverWebSocketPath: string = '/http_over_websocket';
 
+function stopVmHandler(request: http.ServerRequest, response: http.ServerResponse) {
+  if ('POST' != request.method) {
+    return;
+  }
+  try {
+    let vminfo = info.getVmInfo();
+    childProcess.execSync(
+      'gcloud compute instances stop ' + vminfo.vm_name +
+         ' --project ' + vminfo.vm_project + ' --zone ' + vminfo.vm_zone,
+      {env: process.env});
```

Comment:
After the gcloud command, we should run 'sudo shutdown -P now' to make sure the VM shuts itself off.